src/pages/experiments/[experimentId]/data-exploration/components/heatmap/HeatmapPlot.jsx

```diff
@@ -121,83 +122,148 @@ const HeatmapPlot = (props) => {
 
-  const downsample = (groupBy) => {
+  // Defines sorting order according to which group the cell belongs to
+  const positionInGrouping = (cellSetNames, cell) => {
+    let position = 0;
+
+    for (const { key: cellSetName } of cellSetNames) {
+      if (cellsInSet(cellSetName).has(cell)) {
+        return position;
+      }
+
+      position += 1;
+    }
+
+    // If not in grouping set it in the end
+    return position + 1;
+  };
+
+  const cellsInSet = (cellSetName) => {
+    return properties[cellSetName].cellIds;
+  };
+
+  ///////////////////////////////////////////////////////////////////////////////////////////
+  // Returns a set of all cells relevant to the experiment (so the ones belonging to one or more of the enabled sets and not hidden).
+  const getAllRelevantCellIds = (arrayOfcellSetNamesByGroupByKey) => {
     // Find all hidden cells.
     const hiddenCellIds = union(Array.from(hidden), properties);
 
-    // Get how many cells we are sampling.
-    let total = 0;
+    let allCellsInSets = new Set();
 
-    // Create an object for storing the cells grouped by `groupBy`.
-    const groupedCells = {};
+    Object.values(arrayOfcellSetNamesByGroupByKey).forEach((arrayOfSetNames) => {
+      // Add each set of cells into 
+      arrayOfSetNames.forEach((setNameObj) => {
+        const cellSet = cellsInSet(setNameObj.key);
+        // Union of allCellsInSets and cellSet
+        allCellsInSets = new Set([...allCellsInSets, ...cellSet]);
+      });
+    });
 
-    // Find the `groupBy` root node.
-    const rootNode = hierarchy.filter((clusters) => clusters.key === groupBy);
+    // Only work with non-hidden cells.
+    const shownCells = new Set([...allCellsInSets].filter(x => !hiddenCellIds.has(x)));
 
-    if (!rootNode.length) {
-      return [];
+    return shownCells;
+  }
+
+  const getCellSetNamesDict = (rootNodes) => {
+    const cellSetNamesByGroupByKey = {};
+    rootNodes.forEach((currentRootNode) => cellSetNamesByGroupByKey[currentRootNode.key] = currentRootNode.children);
+
+    return cellSetNamesByGroupByKey
+  };
+
+  const downsampled = (cellIds) => {
+    // If sample is within maxCells's limit then return original array
+    if (cellIds.length <= maxCells) {
+      return cellIds;
     }
 
-    const { children } = rootNode[0];
+    let downsampledCellIds = [];
 
-    // Iterate over each child node.
-    children.forEach(({ key }) => {
-      // Only work with non-hidden cells.
-      const shownCells = Array.from(
-        properties[key].cellIds,
-      ).filter(
-        (id) => !hiddenCellIds.has(id),
-      );
+    // Take maxCells amount of cells, picking each one randomly
+    for (let i = 0; i < maxCells; i++) {
+      var randomIndex = Math.floor(Math.random() * cellIds.length);
+      const element = cellIds.splice(randomIndex, 1);
 
-      total += shownCells.length;
+      downsampledCellIds.push(element[0]);
+    }
 
-      groupedCells[key] = shownCells;
-    });
+    return downsampledCellIds;
```

Comment:
You can probably use [_.sampleSize](https://lodash.com/docs/4.17.15#sampleSize) in place of this custom function.